utils/mil_tools/nodes/online_bagger.py

```diff
@@ -319,78 +313,146 @@ def set_bag_directory(self, filename=''):
 
     def start_bagging(self, req):
         """
         Dump all data in dictionary to bags, temporarily stops streaming
         during the bagging process, resumes streaming when over.
         If bagging is already false because of an active call to this service
         """
+        result = BagResult()
         if self.streaming is False:
-            status = 'Bag Request came in while bagging, priority given to prior request'
-            rospy.logwarn(status)
-            return status
-
-        self.streaming = False
-        self.set_bag_directory(req.bag_name)
-
-        self.bagger_status = 'Total Message Count: {} \nFailed Requested Topics: \n'.format(
-            self.get_total_message_count())
-        requested_seconds = req.bag_time
-
-        selected_topics = req.topics.split()
-        for topic, (time, subscribed) in self.subscriber_list.items():
-            if not subscribed:
-                self.bagger_status = self.bagger_status + \
-                    'topic: {}\n'.format(topic)
-                continue
-            # Exclude topics that aren't in topics service argument
-            # If topics argument is empty string, include all topics
-            if len(selected_topics) > 0 and topic not in selected_topics:
-                continue
-            if len(self.topic_messages[topic]) == 0:
-                self.bagger_status = self.bagger_status + \
-                    'topic: {}\n'.format(topic)
-                continue
-
-            rospy.logdebug('message count: {:6}, topic: {}'.format(
-                self.get_topic_message_count(topic), topic))
-
-            # if no number is provided or a zero is given, bag all messages
-            types = (0, 0.0, None)
-            if req.bag_time in types:
-                bag_index = 0
-                self.bag_report = 'All messages were bagged'
-
-            # get time index the normal way
-            else:
-                # bag index will return the appropriate index to begin bagging at
-                # in order to have the requested number of seconds
-                bag_index = self.get_time_index(topic, requested_seconds)
-            rospy.logdebug('topic: {}, bag_index: {}'.format(topic, bag_index))
-            messages = 0  # message number in a given topic
-            for msgs in self.topic_messages[topic][bag_index:]:
-                messages = messages + 1
-                self.bag.write(topic, msgs[1], t=msgs[0])
-                if messages % 100 == 0:  # log every 100th topic, type and time
-                    rospy.logdebug('topic: %s, message type: %s, time: %s',
-                                   topic, type(msgs[1]), type(msgs[0]))
-
-            # empty deque when done writing to bag
-            self.topic_messages[topic].clear()
-
-        self.bag.close()
-        rospy.loginfo('bagging finished!')
+            result.status = 'Bag Request came in while bagging, priority given to prior request'
```

Comment:
This has me wondering how hard dumping separate bags from the queue would be. I don't expect this to be implemented in this pull request, but hear me out:

* When a dump request comes in, it is marked with a `dump_start` and `dump_end` index in the current queue (for the case of one bag at a time, this would just be the beginning and end of the queue)
* When a bag is being dumped, the deletion of old queue entries is halted (but new entries are still being added) and a `deletion_head` is created that moves forward in the queue and marks how many values would have been deleted if no bag was being dumped
* The bag dumped for the initial request above would stop dumping once it reached the `dump_end` marker
* If a second dump request is received while the first one is still running, the `dump_start` index will be set to the current index of the `deletion_head` (that is where the queue would have started if the queue deletion was not paused for the other bag)
* Once a bag finishes dumping, any items at an index before the `deletion_head` and any existing `dump_start` markers is deleted from the queue
* Once there are no bags being dumped, the deletion head can be deleted and the normal rolling queue can be resumed

Basically, this would allow us to dump multiple bags at the same time while still maintaining a rolling queue of messages (for the most part).